Compile PHP files from CLI

akerouanton opened the pull request:
The only compiler available as far as now was the PhpServerCompiler. I'm
not sure why, but it was taking way too much time on our current
project (> 1h).

I added a CliCompiler that just do the same as the former compiler, but
from the CLI. It now takes less than 5 seconds to compile ~5000 files.
Moreover, with this new compiler, we don't need the socket extension
anymore.

The default compiler is still PhpServerCompiler, but the CliCompiler can
be used by specifying --compiler option.

akerouanton committed:
Introduce a new CliCompiler

The only compiler available as far as now was the PhpServerCompiler. I'm
not sure why, but it was taking way too much time on our current
project (> 1h).

This commit adds a CliCompiler that just do the same as the former
compiler, but from the CLI. It takes less than 5 seconds to compile
~5000 files.

Unfortunately, as pointed out by jderusse in his review, not all files
can be compiled in the same process as composer because some of these
files are already loaded in memory, thus compiling them would cause:

>  "PHP Fatal error: Cannot declare class Foo, because the name is
>  already in use in /path/Bar.php on line 3"

jderusse commented on composer.json:
```diff
@@ -4,8 +4,10 @@
+    "ext-Zend-OPcache": "*"
+  },
+  "suggest": {
+    "ext-sockets": "Needed to use the PhpServerCompiler"
```
should not be a suggestion if `php_server` is the default compiler

jderusse commented on src/Console/WarmupCommand.php:
```diff
@@ -6,19 +6,31 @@
 
 class WarmupCommand extends BaseCommand
 {
+    const COMPILER_PHP_SERVER = 'php_server';
```
I'd rather use `sandbox` as the use of php server is a technical detail of the current implementation

jderusse commented on src/Console/WarmupCommand.php:
```diff
@@ -59,15 +71,33 @@ function ($extra) {
+
+        if ($compilerOption === self::COMPILER_PHP_SERVER) {
+            $compiler = new PhpServerCompiler();
+        } elseif ($compilerOption !== self::COMPILER_CLI) {
```
a `switch` would have been more readable IMHO

jderusse reviewed (commented):
The purpose of the server was to "sandbox" the compilation, and avoid 2 side effects:
1) When a class is used in several places (which is the case for all dependencies embedded in composer and this plugin) the call to `opcache_compile_file` would trigger a "PHP Fatal error:  Cannot declare class Foo, because the name is already in use in /path/Bar.php on line 3"

2) The compiled file will be loaded, and the autoloader won't be triggered for class loaded. Meaning that if Composer or this plugin try to load a class after the compilation phase (display error/log, result) the instanciate class may not be the expected one.  

I don't know how to fix the 2nd issue (which may be negligible ?) but the first one, but for the 1st issue, what could be done instead of asking to the user one compiler or the other is to wrap the 2 implementations in a FallbackCompiler

```php
class FallackCompiler implements CompilerInterface
{
  public function __construct(iterable $compilers)
  {
    $this->compilers = $compilers
  }
  public function compile(string $file)
  {
   foreach ($this->compilers as $compiler) {
     try {
       return $compiler->compile($file);
     } catch (\Throwable $e) {
     }
    throw new \Exception('failed to compile file');
  }
}
```
to use
```
$compiler = new FallackCompiler([new FileCompiler(), new ServerCompiler()]);
```

akerouanton committed:
Introduce a FallbackCompiler

This new compiler takes a set of compilers and try to compile files
until one of the compiler succeed. In this way, we can use the newly
introduced CliCompiler for most of the files and fallback to the former
PhpServerCompiler when it fails.

akerouanton commented:
> The purpose of the server was to "sandbox" the compilation

I was pretty sure there was a good reason to do it like that in the first place but I wasn't sure why actually :stuck_out_tongue: Indeed I saw the 1st error in our build logs after submitting this PR. For the 2nd error, I shall admit I'm not sure to get what you mean :thinking: 

I implemented the `FallbackCompiler` and removed the `--compiler` option as suggested :wink: 

I also tried to implement a `SubprocessCompiler` that would just do something like `php -r "opcache_compile_file($file);"` but I couldn't make it work. I prefer to move on. But if I've some free time in the upcoming weeks, I'll try to give it a shot a second time.

jderusse commented:
Thank you @NiR-

akerouanton commented:
You welcome :) Any chance these changes got released soon? :)

jderusse commented:
done

staabm commented:
just a side-note, because we were trapped by similar problems:

generating files from the CLI and using them from a different process (e.g. FPM or Apache) will not work out of the box in Ubuntu18+ because of systemd and the PrivateTmp feature, which sandboxes the files within /tmp

jderusse commented:
`/tmp/` cqn be configured right ?

jderusse commented:
`/tmp/` can be configured right ?

staabm commented:
> `/tmp/` can be configured right ?

correct, this can be configured. my point: as is it will not work with default settings.

jderusse commented:
Documentation PR welcome :-)
